### PR TITLE
Wipes plaintext email for students from Poste2.

### PR DESCRIPTION
**Pull request by ashercodeorg**
Adds significant test coverage to existing methods.

### CODE REVIEW

File: lib/cdo/poste.rb [hunk 1]
```diff
@@ -170,12 +171,11 @@ def self.ensure_recipient(email, params={})
...
-    contacts = POSTE_DB[:contacts]
-
-    contact = contacts.where(hashed_email: hashed_email).first
+    contact = POSTE_DB[:contacts].where(hashed_email: hashed_email).first
     unless contact
-      id = contacts.insert({}.tap do |contact|
-        contact[:email] = email
+      sanitized_email = Poste.dashboard_student?(hashed_email) ? '' : email
+      id = POSTE_DB[:contacts].insert({}.tap do |contact|
+        contact[:email] = sanitized_email
```
Comment (aoby) [inline lib/cdo/poste.rb:178]: Not important: I know you didn't write this code, and it's mostly extraneous to this PR, but any idea why it's using `{}.tap` rather than just creating a hash inline? 

That seems a strange and unnecessary pattern. The only reason I can see is the optional name field, but since this is inserting in the DB wouldn't a hash with `name: nil` resolve to the same thing as as a hash with no `name` key?

It can be converted to:

``` ruby
id = POSTE_DB[:contacts].insert({
  email: sanitized_email,
  hashed_email: hashed_email,
  name: name,
  #...
})
```

File: shared/test/test_poste.rb [hunk 2]
```diff
@@ -1,57 +1,69 @@
...
-    )
-
-    Poste.unsubscribe(email, hashed_email, {ip_address: '1.2.3.4'})
+    Poste.unsubscribe(
+      STUDENT_EMAIL, STUDENT_EMAIL_HASH, {ip_address: '1.2.3.4'})
 
     assert POSTE_DB[:contacts].
-      where(hashed_email: hashed_email).first[:unsubscribed_at]
-    assert_nil POSTE_DB[:contacts].where(email: email).first
+      where(hashed_email: STUDENT_EMAIL_HASH).first[:unsubscribed_at]
```
Comment (aoby) [inline shared/test/test_poste.rb]: Does this work? I don't see where the student is added to code_studio.

File: shared/test/test_poste.rb [hunk 3]
```diff
@@ -1,57 +1,69 @@
...
+    assert_nil POSTE_DB[:contacts].where(email: STUDENT_EMAIL).first
```
Comment (aoby) [inline shared/test/test_poste.rb:59]: Does this work? I don't see where the student is added to code_studio.

File: shared/test/test_poste.rb [hunk 4]
```diff
@@ -1,57 +1,69 @@
...
-      username: 'code studio teacher',
-      user_type: 'teacher',
-      birthday: '1990-01-02'
-    )
+    assert !Poste.dashboard_student?(hashed_email)
+  end
 
-    Poste.unsubscribe(email, hashed_email, {ip_address: '1.2.3.4'})
+  def test_dashboard_student_for_student
+    assert Poste.dashboard_student?(STUDENT_EMAIL_HASH)
```
Comment (aoby) [inline shared/test/test_poste.rb:79]: same as above. Am I missing where the `STUDENT_EMAIL_HASH` is added to dashboard?

File: shared/test/test_poste.rb [hunk 5]
```diff
@@ -192,4 +204,104 @@ def test_multipart_mail
...
+      created_ip: '1.2.3.4',
+      updated_at: DateTime.now,
+      updated_ip: '1.2.3.4'
+    })
+
+    recipient = Poste2.ensure_recipient(email, {ip_address: '5.6.7.8'})
+    # Intentional or not, the IP address returned is the IP address passed in.
+    assert_equal '5.6.7.8', recipient[:ip_address]
+    assert POSTE_DB[:contacts].where(hashed_email: hashed_email).first
+    # Intentional or not, the IP address is not updated in the DB.
```
Comment (aoby) [inline shared/test/test_poste.rb:320]: :(

**Comment by aoby** [comment]
Nice new test coverage!

It looks like some setup is missing (or I am missing something). See above comments.

File: shared/test/test_poste.rb (hunk 3 above)
Comment (ashercodeorg) [inline shared/test/test_poste.rb:59]: Oops, I must have deleted the `setup` method when dealing with the merge conflict. Let me try again.

File: lib/cdo/poste.rb (hunk 1 above)
Comment (ashercodeorg) [inline lib/cdo/poste.rb:178]: Nope. And I'm happy to "fix" in a followup PR.

**Comment by aoby** [comment]
LGTM with the setup in the 2nd commit, presuming tests pass.
